### ISSUE TITLE
podman pull: fails if $XDG_RUNTIME_DIR does not exist

### ISSUE THREAD

**Issue by edsantiago** [projectatomic/libpod#659]
[ Related to #427 ]

If $XDG_RUNTIME_DIR does not exist, `podman pull` fails with an unhelpful message:
```
# podman pull fedora
Trying to pull docker.io/fedora:latest...Failed
Trying to pull registry.fedoraproject.org/fedora:latest...Failed
Trying to pull registry.access.redhat.com/fedora:latest...Failed
error pulling image "fedora": unable to pull fedora
# echo $XDG_RUNTIME_DIR 
/run/user/0
# ls -l $XDG_RUNTIME_DIR
ls: cannot access '/run/user/0': No such file or directory
# env -u XDG_RUNTIME_DIR podman pull fedora
Trying to pull docker.io/fedora:latest...Getting image source signatures
...success...
```
I am using `mosh` to log in; mosh creates an initial connection using `ssh`, which creates that directory, but the ssh connection is short-lived and the directory soon disappears.

**Comment by mheon** [comment]
It's fine for this to be a hard failure for `podman login`, but I expect `podman pull` to just warn and continue if we can't find login credentials.

**Comment by umohnani8** [comment]
@edsantiago I am assuming you are pulling a private image from your docker hub?

**Comment by edsantiago** [comment]
@umohnani8 no, see comment 0 above: with XDG_RUNTIME_DIR cleared, it fetches `docker.io/fedora:latest` (first in /etc/containers/registries.conf). Or, for rhel7, the first registry that works:
```
# env -u XDG_RUNTIME_DIR podman pull rhel7
Trying to pull docker.io/rhel7:latest...Failed
Trying to pull registry.fedoraproject.org/rhel7:latest...Failed
Trying to pull registry.access.redhat.com/rhel7:latest...Getting image source signatures
Skipping fetch of repeat blob sha256:d1fe25896eb5cbceebe612f6f820f433997d6748700d05f2029e8e077c02fd16
Skipping fetch of repeat blob sha256:001d79f6847029de668d289c5b602d004b2bf1691d29f31c643481e0821f03d2
Copying config sha256:fd1ba0b398a82d56900bb798c8b099fbe3166bc49e2c5e947f7973cd38ff1a90
 6.13 KB / 6.13 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
fd1ba0b398a82d56900bb798c8b099fbe3166bc49e2c5e947f7973cd38ff1a90
```
You should be able to reproduce this via:
```
# env XDG_RUNTIME_DIR=/nonexistent podman pull fedora (or centos or rhel7 or anything not already present)
```

**Comment by umohnani8** [comment]
But the credentials shouldn't matter if you are pulling a public image. It works for me:
```
[root@dhcp-10-20-1-25 libpod]# echo $XDG_RUNTIME_DIR

[root@dhcp-10-20-1-25 libpod]# podman pull fedora
Trying to pull docker.io/fedora:latest...Getting image source signatures
Copying blob sha256:2176639d844bbe1386912e1d9952cebdb8249923a16691025cf693963f8aec53
 77.54 MB / 77.54 MB [======================================================] 3s
Copying config sha256:9110ae7f579f35ee0c3938696f23fe0f5fbe641738ea52eb83c2df7e9995fa17
 2.29 KB / 2.29 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
9110ae7f579f35ee0c3938696f23fe0f5fbe641738ea52eb83c2df7e9995fa17
```

**Comment by edsantiago** [comment]
Try with XDG_RUNTIME_DIR=/nonexistent

**Comment by umohnani8** [comment]
It still works for me
```
[root@dhcp-10-20-1-25 buildah]# XDG_RUNTIME_DIR=/nonexistent
[root@dhcp-10-20-1-25 buildah]# echo $XDG_RUNTIME_DIR
/nonexistent
[root@dhcp-10-20-1-25 buildah]# podman pull fedora
Trying to pull docker.io/fedora:latest...Getting image source signatures
Copying blob sha256:2176639d844bbe1386912e1d9952cebdb8249923a16691025cf693963f8aec53
 77.54 MB / 77.54 MB [======================================================] 3s
Copying config sha256:9110ae7f579f35ee0c3938696f23fe0f5fbe641738ea52eb83c2df7e9995fa17
 2.29 KB / 2.29 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
9110ae7f579f35ee0c3938696f23fe0f5fbe641738ea52eb83c2df7e9995fa17
```

**Comment by mheon** [comment]
I can confirm that the reproducer @edsantiago pasted fails for me too

**Comment by cgwalters** [comment]
> but the ssh connection is short-lived and the directory soon disappears.

IMO that's a mosh flaw; it should talk to `logind` and have it hold the original session open the for as long as it makes the login available; that'd require some tricky work with PAM+ssh probably though.

**Comment by edsantiago** [comment]
>IMO that's a mosh flaw; it should talk to logind

As a UNIX old-timer I resent that argument, but ultimately agree that yes, tools have to adapt to this modern systemd world. Still, it's also a bug that `podman pull` fails in the case above.

**Comment by umohnani8** [comment]
@rhatdan so if XDG_RUNTIME_DIR is nonexistent, should we error out or log a warning and move on? What is the approach we want here?

**Comment by mheon** [comment]
@umohnani8 I really do not want to error out on this in `pull` and similar cases. We should proceed (but warn), and if the pull needs credentials and we don't have them because no XDG_RUNTIME_DIR, we error then.

**Comment by umohnani8** [comment]
Fixed by #451
